src/include/homestore/btree/btree_kv.hpp

```diff
@@ -45,37 +44,43 @@ class BtreeKey {
     // Deleting copy constructor forces the derived class to define its own copy constructor
     // BtreeKey(const BtreeKey& other) = delete;
     // BtreeKey(const sisl::blob& b) = delete;
-    BtreeKey(const BtreeKey& other) = default;
+    BtreeKey(BtreeKey const& other) = default;
     virtual ~BtreeKey() = default;
 
-    virtual BtreeKey& operator=(const BtreeKey& other) {
-        clone(other);
-        return *this;
-    };
-
-    virtual void clone(const BtreeKey& other) = 0;
-    virtual int compare(const BtreeKey& other) const = 0;
+    virtual int compare(BtreeKey const& other) const = 0;
 
     virtual sisl::blob serialize() const = 0;
     virtual uint32_t serialized_size() const = 0;
-    virtual void deserialize(const sisl::blob& b, bool copy) = 0;
+    virtual void deserialize(sisl::blob const& b, bool copy) = 0;
 
     virtual std::string to_string() const = 0;
-    virtual bool is_extent_key() const { return false; }
+    virtual bool is_interval_key() const { return false; }
```

Comment:
I know it is from the legacy code, a little odd to me that base class has knowledge about subclass. 
it looks like same goal can be achieved by try to dynamic cast from a base class pointer. Is there any problem if we remove this api?

Reply:
No it is not that base class is aware of derived class, but by default they are not interval key. Only BtreeIntervalKey can claim that. If this is not there, then we need to define additional BtreeNonIntervalKey which does only this method. We can't do it with dynamic_cast either, because btree main library will take key and call `is_interval_key()`,